toolchain/sem_ir/ids.h

```diff
@@ -137,13 +149,22 @@ struct ConstantId : public IdBase, public Printable<ConstantId> {
   }
 
  private:
-  // ConstantIds don't have an invalid state.
-  using IdBase::is_valid;
+  // The offset of InstId indices to ConstantId indices.
+  static constexpr int32_t IndexOffset = 2;
+  static constexpr int32_t NotConstantIndex = IndexOffset + InvalidIndex;
```

Comment:
It's not obvious to me what the pattern is here (eg, how this would be extended if we added more special indexes). I can see that this is right because I know that `InvalidIndex == -1`, but if we had a second special value or a third, I'm not clear on what calculation we'd use.

Maybe it'd be more consistent to use all negative values for special indices, like other `Id` types:

```suggestion
  static constexpr int32_t NotConstantIndex = InvalidIndex - 1;
  // The offset of InstId indices to ConstantId indices.
  static constexpr int32_t IndexOffset = -NotConstantIndex + 1;
```

(I think keeping the offset for both the positive and negative cases is probably still a good thing, so that we quickly catch cases where an `InstId` is confused for a `ConstantId`, but I'd be OK with having the offset only for the negative cases if you'd prefer.)

Reply:
I was trying to have the `inst_id` calculation trivial, are you intending that there should be control flow? i.e., instead of:

```
return InstId(std::abs(index) - IndexOffset);
```

(where `std::abs` might be a `neg` instruction)

We have multiple branches, such as:

```
if (index >= 0) {
  return InstId(index);
} else if (!is_valid() || *this == NotConstant) {
  return InstId::Invalid;
} else {
  return InstId(-index - IndexOffset);
}
```

Reply:
No, I was intending that the `inst_id` calculation would remain the same -- we would keep the offset for both the positive and negative cases.

Reply:
Okay, but the `inst_id` calculation still needs to change to handle invalid _somehow_. Assuming you're wanting this, I've added a `CARBON_CHECK` that it's only used for valid constants.